src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value_impl.hpp

```diff
@@ -0,0 +1,399 @@
+/**
+ * @file discrete_hilbert_value.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Defintion of the DiscreteHilbertValue class, a class that calculates
+ * the ordering of points using the Hilbert curve.
+ */
+#ifndef MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+#define MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+
+#include "discrete_hilbert_value.hpp"
+
+namespace mlpack {
+namespace tree /** Trees and tree-building procedures. */ {
+
+inline DiscreteHilbertValue::DiscreteHilbertValue() :
+    dataset(new arma::Mat<uint64_t>()),
+    ownsDataset(true),
+    localDataset(new std::list<arma::Col<uint64_t>>()),
+    largestValue(localDataset->end())
+{
+
+};
+
+inline DiscreteHilbertValue::~DiscreteHilbertValue()
+{
+  delete localDataset;
+  if(ownsDataset)
+    delete dataset;
+};
+
+template<typename TreeType>
+DiscreteHilbertValue::DiscreteHilbertValue(const TreeType *tree) :
+    dataset(tree->Parent() ?
+          tree->Parent()->AuxiliaryInfo().LargestHilbertValue().Dataset() :
+          new arma::Mat<uint64_t>(tree->Dataset().n_rows,
+                                  tree->Dataset().n_cols)),
+    ownsDataset(!tree->Parent()),
+    localDataset(new std::list<arma::Col<uint64_t>>()),
+    largestValue(localDataset->end())
+{
+  typedef typename TreeType::ElemType ElemType;
+  // Calculate the Hilbert value for all points
+  if(!tree->Parent())
+  {
+    for(size_t i = 0; i < tree->Dataset().n_cols; i++)
+      dataset->col(i) = CalculateValue((arma::Col<ElemType>)tree->Dataset().col(i));
+  }
+}
+
+template<typename TreeType>
+DiscreteHilbertValue::DiscreteHilbertValue(const TreeType &other) :
+    dataset(other.AuxiliaryInfo().LargestHilbertValue().Dataset()),
+    ownsDataset(false),
+    localDataset(new std::list<arma::Col<uint64_t>>()),
+    largestValue(other.AuxiliaryInfo().LargestHilbertValue().LargestValue())
+{
+  if(other.IsLeaf())
+  {
+    std::list<arma::Col<uint64_t>> *otherDataset =
+                     other.AuxiliaryInfo().LargestHilbertValue().LocalDataset();
+    for(std::list<arma::Col<uint64_t>>::iterator it = otherDataset->begin(); it != otherDataset->end(); it++)
+    {
+      localDataset->push_back(*it);
+    }
+    largestValue = localDataset->end();
+    if(otherDataset->size() > 0)
+      largestValue--;
+  }
+}
+
+template<typename ElemType>
+arma::Col<uint64_t> DiscreteHilbertValue::
+CalculateValue(const arma::Col<ElemType> &pt)
+{
+  arma::Col<uint64_t> res(pt.n_rows);
+  constexpr int order = 64;   // The number of bits that we can store
```

Comment:
Hm, is this the right order if ElemType is a float?  Or should we be using `uint32_t` in that case?  It would not be too hard to do a little template metaprogramming to figure out the right type to use (`uint32_t` or `uint64_t`), but I am not 100% sure if the order should be 32 if ElemType is a float, you would know better than me.